gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpHeadersUtils.java

```diff
@@ -0,0 +1,103 @@
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.httpjson;
+
+import com.google.api.client.http.HttpHeaders;
+import com.google.api.client.util.ClassInfo;
+import com.google.api.client.util.FieldInfo;
+import com.google.api.core.InternalApi;
+import com.google.common.collect.ImmutableList;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class is for internal use only and is public for technical reasons.
+ */
+@InternalApi
+public class HttpHeadersUtils {
+  public static String getUserAgentValue(Map<String, String> headersMap) {
+    for (Map.Entry<String, String> entry : headersMap.entrySet()) {
+      if ("user-agent".equals(entry.getKey().toLowerCase())) {
+        return entry.getValue();
+      }
+    }
+    return null;
+  }
+
+  public static HttpHeaders setHeaders(HttpHeaders headers, Map<String, String> headersMap) {
+    for (Map.Entry<String, String> entry : headersMap.entrySet()) {
+      setHeader(headers, entry.getKey(), entry.getValue());
+    }
+    return headers;
+  }
+
+  public static HttpHeaders setHeader(HttpHeaders headers, String key, String value) {
+    ClassInfo classInfo = headers.getClassInfo();
+    FieldInfo fieldInfo = classInfo.getFieldInfo(key);
+    Object headerValue;
+    if (fieldInfo == null) {
+      headerValue = value;
+    } else {
+      if (List.class.isAssignableFrom(fieldInfo.getType())) {
+        Type genericType = fieldInfo.getGenericType();
+        if (genericType instanceof ParameterizedType) {
+          ParameterizedType parameterizedType = (ParameterizedType) genericType;
+          // parameterizedType is a list, so is guaranteed to have at least one type argument
+          Type listElementType = parameterizedType.getActualTypeArguments()[0];
+          if (listElementType instanceof Class) {
+            headerValue = ImmutableList.of(fromString((Class) listElementType, value));
+          } else {
+            throw new IllegalArgumentException();
```

Comment:
Can you give more details in this exception?

Reply:
Added description.